From 46e804441971e2678ff403c47bfe6046e20168aa Mon Sep 17 00:00:00 2001
From: Kent Tamura <tkent@chromium.org>
Date: Fri, 21 Jun 2024 08:40:43 +0000
Subject: [PATCH] Reland "Layout: Fix an InlineCursor crash with a multicol container"

This is a reland of commit 6ea422dcf50a8b03fd5047af2952f30c6583d731

The original commit was reverted by a bot mistakenly.

Original change's description:
> Layout: Fix an InlineCursor crash with a multicol container
>
> InlineCursor::MoveToVisualFirstOrLastForCulledInline() had a bug that
> it tried to move to an item in a different root.
>
> The function has a loop by MoveToNextForSameLayoutObject(), which can
> be moved to another root if the container is block-fragmented.  This
> CL fixes it by adjusting the root before MoveTo().
>
> Bug: 343749045
> Change-Id: I856253e8e61163a7705676704ea0c8451096acb5
> Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/5644191
> Auto-Submit: Kent Tamura <tkent@chromium.org>
> Commit-Queue: Koji Ishii <kojii@chromium.org>
> Reviewed-by: Koji Ishii <kojii@chromium.org>
> Cr-Commit-Position: refs/heads/main@{#1317762}

Bug: 343749045
Change-Id: Ifa0890d6bc5fdf9f0f27ecd72efb19f9d2321449
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/5645473
Bot-Commit: Rubber Stamper <rubber-stamper@appspot.gserviceaccount.com>
Auto-Submit: Kent Tamura <tkent@chromium.org>
Commit-Queue: Rubber Stamper <rubber-stamper@appspot.gserviceaccount.com>
Cr-Commit-Position: refs/heads/main@{#1317829}
---

--- a/third_party/blink/renderer/core/layout/inline/inline_cursor.cc
+++ b/third_party/blink/renderer/core/layout/inline/inline_cursor.cc
@@ -1732,6 +1732,7 @@
 void InlineCursor::MoveToVisualFirstOrLastForCulledInline(bool last) {
   InlineCursorPosition found_position;
   std::optional<size_t> found_index;
+  wtf_size_t found_fragment_index = 0;
 
   // Iterate through the remaining fragments to find the lowest/greatest index.
   for (; Current(); MoveToNextForSameLayoutObject()) {
@@ -1742,6 +1743,7 @@
         (!last && index < *found_index)) {
       found_position = Current();
       found_index = index;
+      found_fragment_index = fragment_index_;
 
       // Break if there cannot be any fragment lower/greater than this one.
       if ((last && index == fragment_items_->Size() - 1) ||
@@ -1751,6 +1753,13 @@
   }
 
   DCHECK(found_position);
+  if (RuntimeEnabledFeatures::InlineCursorMultiColFixEnabled() &&
+      fragment_index_ > found_fragment_index) {
+    while (fragment_index_ > found_fragment_index) {
+      DecrementFragmentIndex();
+    }
+    CHECK(TrySetRootFragmentItems());
+  }
   MoveTo(found_position);
 }
 
--- a/third_party/blink/renderer/platform/runtime_enabled_features.json5
+++ b/third_party/blink/renderer/platform/runtime_enabled_features.json5
@@ -2128,6 +2128,11 @@
       status: "stable",
     },
     {
+      // crbug.com/343749045
+      name: "InlineCursorMultiColFix",
+      status: "stable",
+    },
+    {
       // If enabled, and the inner html parser is unable to successfully
       // parse, it will log histograms of why it failed. The logging is
       // non-trivial.
--- /dev/null
+++ b/third_party/blink/web_tests/editing/caret/multicol-rtl-crash.html
@@ -0,0 +1,34 @@
+<!DOCTYPE html>
+<style type="text/css">
+:last-child {
+  break-inside:auto;
+  column-count:112 !important;
+  stroke-width:inherit;
+}
+*:only-of-type {
+  display:inline;
+}
+</style><script defer="defer" type="text/javascript">
+function event_handler_1307_readystatechange() {
+  var oSelection=window.getSelection();
+  oSelection.modify('move', 'forward', 'paragraphboundary')
+  if (!oSelection.rangeCount) {
+    document.execCommand("SelectAll")
+  }
+}
+document.addEventListener("readystatechange", event_handler_1307_readystatechange);
+</script>
+<body dir="rtl">
+<table>
+<caption>
+<textarea>
+</textarea>
+AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA
+<input>
+<script src="../../resources/testharness.js"></script>
+<script src="../../resources/testharnessreport.js"></script>
+<script>
+test(() => {
+  document.querySelector('style').remove();
+}, 'crbug.com/343749045. Pass if no crashes');
+</script>
